### PR TITLE
Wire up stats commands for daemonsets (#2006)

### PR DESCRIPTION
**Pull request by zknill**
DaemonSet stats are not currently shown in the cli stat command, web ui
or grafana dashboard. This commit adds daemonset support for stat.

Update stat command's help message to reference daemonsets.
Update the public-api to support stats for daemonsets.
Add tests for stat summary and api.

Add daemonset get/list/watch permissions to the linkerd-controller
cluster role that's created using the install command.
Update golden expectation test files for install command
yaml manifest output.

Update web UI with daemonsets
Update navigation, overview and pages to list daemonsets and the pods
associated to them.
Add daemonset paths to server, and ui apps.

Add grafana dashboard for daemonsets; a clone of the deployment
dashboard.

Fixes of #2006

Signed-off-by: Zak Knill <zrjknill@gmail.com>

### CODE REVIEW

**Comment by zknill** [comment]
# Validation

To test, took the example project `curl https://run.linkerd.io/emojivoto.yml > emojivoto.yaml `, and updated the Deployments for `web`, `voting` and `emoji` to be DaemonSets

## cli

linkerd stat all -n emojivoto
linkerd stat daemonset -n emojivoto
linkerd stat ds -n emojivoto

linkerd stat daemonset/emoji -n emojivoto
linkerd stat ds/web -n emojivoto

linkerd stat daemonsets --to ds/emoji -n emojivoto
linkerd stat daemonsets --from ds/web -n emojivoto


```
$ linkerd stat ds -n emojivoto
NAME     MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TLS
emoji       1/1   100.00%   1.9rps           1ms           3ms           3ms    0%
voting      1/1    79.66%   1.0rps           1ms           3ms           4ms    0%
web         1/1    89.57%   1.9rps          10ms          37ms          40ms    0%
```

## web UI

Pages checked;
- /overview
- /daemonsets
- /namespaces/:namespace/daemonsets/:daemonset

## grafana

Checked dashboard, and links from above pages of the web ui

**Comment by wmorgan** [comment]
Awesome, thanks for submitting this @zknill!

**Comment by siggy** [comment]
@zknill awesome! would you mind sharing your modified emojivoto/daemonsets config? bonus points for submitting it as a PR to https://github.com/BuoyantIO/emojivoto ;)

File: controller/k8s/api.go [hunk 1]
```diff
@@ -33,6 +35,7 @@ type APIResource int
 // These constants enumerate Kubernetes resource types.
 const (
 	CM APIResource = iota
+	Daemonset
```
Comment (siggy) [inline controller/k8s/api.go]: mind naming this `DS`?

File: controller/k8s/api.go [hunk 2]
```diff
@@ -47,15 +50,16 @@ const (
...
-	deploy   appinformers.DeploymentInformer
-	endpoint coreinformers.EndpointsInformer
-	mwc      arinformers.MutatingWebhookConfigurationInformer
-	pod      coreinformers.PodInformer
-	rc       coreinformers.ReplicationControllerInformer
-	rs       appinformers.ReplicaSetInformer
-	sp       spinformers.ServiceProfileInformer
-	svc      coreinformers.ServiceInformer
+	cm        coreinformers.ConfigMapInformer
+	daemonset appv1informers.DaemonSetInformer
```
Comment (siggy) [inline controller/k8s/api.go]: `ds`

File: controller/api/public/stat_summary_test.go [hunk 3]
```diff
@@ -209,6 +209,82 @@ status:
...
 	})
 
+	t.Run("Successfully performs a query based on resource type DaemonSet", func(t *testing.T) {
+		expectations := []statSumExpected{
+			statSumExpected{
+				expectedStatRPC: expectedStatRPC{
+					err: nil,
+					k8sConfigs: []string{`
+apiVersion: apps/v1
+kind: DaemonSet
```
Comment (siggy) [inline controller/api/public/stat_summary_test.go]: 🙏

File: cli/cmd/stat.go [hunk 4]
```diff
@@ -59,6 +59,7 @@ func newCmdStat() *cobra.Command {
   Examples:
   * deploy
   * deploy/my-deploy
+  * ds/my-daemonset
```
Comment (siggy) [inline cli/cmd/stat.go]: a few lines below this, add `daemonsets` under `Valid resource types include:`

**Comment by zknill** [comment]
@siggy very happy to open a PR, but not super sure where / how it would fit into that project. It seems more logical for that repo to hold deployments for the services... anyhow, linked below is the daemonset config that I used for testing.

https://gist.github.com/zknill/c2253bad3cc2da0a0b3b10617d2f2bc8

**Comment by grampelberg** [comment]
@zknill just create a secondary yaml named `emojivoto-daemonset.yml` in the root =) If you want *extra* brownie points, a PR on [linkerd/website](https://github.com/linkerd/website) putting the same file at `run.linkerd.io/public/emojivoto-daemonset.yml` would be mind blowing!

File: grafana/dashboards/daemonset.json [hunk 5]
```diff
@@ -0,0 +1,2241 @@
...
+      "6h",
+      "12h",
+      "24h",
+      "2d",
+      "7d",
+      "30d"
+    ]
+  },
+  "timezone": "",
+  "title": "Linkerd Daemonset",
```
Comment (klingerf) [inline grafana/dashboards/daemonset.json]: Can we call this "Linkerd DaemonSet" to match the Kubernetes casing?

**Comment by zknill** [comment]
Sorry for slow response, on holiday at the moment. I will be able to look into things on Sunday / Monday

**Comment by zknill** [comment]
Hi there @klingerf, I intentionally omitted the `top` command support because I thought it out of scope of the original ticket.

I can look at adding it as that experience for users of the UI isn't great.

It might be worth updating the other similar tickets for `StatefulSets` / `Jobs` to be clear that they will also need `top` support?

**Comment by zknill** [comment]
I am looking into top and tap commands next

**Comment by zknill** [comment]
Yes ofc, sorry all!

I'm sure you appreciate it's sometimes hard to find the time to do all the things - I've not forgotten!

**Comment by zknill** [comment]
I think I've got everything; rebase and all

Let me know if there are other tweaks to be made! :pray:

**Comment by zknill** [comment]
@siggy thank you so much for this! (not quite sure how this slipped through the net)

I've picked the commit, and rechecked the UI and stat, top and tap cli comamnds.

Thanks for all your help, I appreciate your time!

**Pull request merged**